src/Chunk.ts

```diff
@@ -441,20 +441,29 @@ export default class Chunk {
 					? '[name].js'
 					: '[name][extname].js'
 				: options.entryFileNames;
-			path = relative(
-				preserveModulesRelativeDir,
-				`${dirname(sanitizedId)}/${renderNamePattern(
-					pattern,
-					'output.entryFileNames',
-					{
-						ext: () => extension.substr(1),
-						extname: () => extension,
-						format: () => options.format as string,
-						name: () => this.getChunkName()
-					},
-					this.getChunkInfo.bind(this)
-				)}`
+			let currentDir = dirname(sanitizedId);
+			const fileName = renderNamePattern(
+				pattern,
+				'output.entryFileNames',
+				{
+					ext: () => extension.substr(1),
+					extname: () => extension,
+					format: () => options.format as string,
+					name: () => this.getChunkName()
+				},
+				this.getChunkInfo.bind(this)
 			);
+			if (options.preserveModulesRoot) {
+				const preserveModulesRoot = resolve(options.preserveModulesRoot);
```

Comment:
There is quite a bit of `resolving` going on here which is not free as `resolve` involves non-trivial file-system operations. So one way to get rid of the first one at least would be to perform this resolution once in `normalizeOutputOptions` so that `options.preserveModulesRoot` is already resolved at this point.